### ISSUE TITLE
unexpected warnings with builtin_dump_struct

### ISSUE THREAD

**Issue by BaoshanPang** [llvm/llvm-project#76084]
```
$ cat ~/tmp/t.c
int printf (const char *restrict, ...)
               __attribute__ ((__format__ (__printf__, 1, 2)));

typedef unsigned char u8;
typedef unsigned char u08;
typedef unsigned short u16;
typedef unsigned int u32;
typedef unsigned long long u64;

struct llu_dump_test {
  u64 a : 4;
  u64 b : 56;
  u64 c : 4;
};

int main() {
  struct llu_dump_test lludump;

  __builtin_dump_struct(&lludump, printf);
  return 0;
}

$ /home/bpang/mywork/llvm_github/llvm-project/build/bin/clang -c ~/tmp/t.c
/home/bpang/tmp/t.c:19:25: warning: format specifies type 'unsigned long long' but the argument has type 'int' [-Wformat]
   19 |   __builtin_dump_struct(&lludump, printf);
      |                         ^
/home/bpang/tmp/t.c:19:3: note: in call to printing function with arguments '("%s%s %s : %zu = %llu\n", "  ", "u64", "a", 4U, (&lludump)->a)' while dumping struct
   19 |   __builtin_dump_struct(&lludump, printf);
      |   ^
/home/bpang/tmp/t.c:19:25: warning: format specifies type 'unsigned long long' but the argument has type 'int' [-Wformat]
   19 |   __builtin_dump_struct(&lludump, printf);
      |                         ^
/home/bpang/tmp/t.c:19:3: note: in call to printing function with arguments '("%s%s %s : %zu = %llu\n", "  ", "u64", "c", 4U, (&lludump)->c)' while dumping struct
   19 |   __builtin_dump_struct(&lludump, printf);
      |   ^
2 warnings generated.

```

**Comment by shafik** [comment]
Confirmed

CC @zyn0217

**Comment by shafik** [comment]
FYI putting code in a separate block is helpful for folks who want to reproduce by cutting and pasting just the code. Ideally providing a godbolt link as well is super helpful e.g.: https://godbolt.org/z/cdjbj5jd9

**Comment by zyn0217** [comment]
This has happened since clang 15. https://cpp2.godbolt.org/z/vs83MKn7T

**Comment by tbaederr** [comment]
This is unrelated to `__builtin_dump_struct`:

```c++
int printf (const char *restrict, ...)
               __attribute__ ((__format__ (__printf__, 1, 2)));

struct llu_dump_test {
  unsigned long a : 3;
  unsigned long b : 32;
};

int main() {
  struct llu_dump_test lludump;

  printf("%lu", lludump.a);
  printf("%lu", lludump.b);

  return 0;
}
```

https://godbolt.org/z/4EsPj61bh

```console
test.cpp:12:17: warning: format specifies type 'unsigned long' but the argument has type 'int' [-Wformat]
   12 |   printf("%lu", lludump.a);
      |           ~~~   ^~~~~~~~~
      |           %d
test.cpp:13:17: warning: format specifies type 'unsigned long' but the argument has type 'unsigned int' [-Wformat]
   13 |   printf("%lu", lludump.b);
      |           ~~~   ^~~~~~~~~
      |           %u
2 warnings generated.
```

Apparently the printf checking doesn't work correctly when bitfields are involved. It seems to pick a type based on the bitfield width.

**Comment by zyn0217** [comment]
> Apparently the printf checking doesn't work correctly when bitfields are involved. It seems to pick a type based on the bitfield width.

I think this should be https://github.com/llvm/llvm-project/issues/75984#issuecomment-1863754778 :-)

**Comment by zyn0217** [comment]
https://github.com/llvm/llvm-project/blob/d5c98e783779c4e6b26b2010b20cd0ab7210ead3/clang/lib/Sema/SemaChecking.cpp#L643-L650

(I guess we should be building an ImplictCastExpr here for bit fields, but seemingly not.)

**Comment by BaoshanPang** [comment]
The warning would be gone with this change:

diff --git a/clang/lib/Sema/SemaChecking.cpp b/clang/lib/Sema/SemaChecking.cpp
index 2e61dafacab0..a74befe8f8b1 100644
--- a/clang/lib/Sema/SemaChecking.cpp
+++ b/clang/lib/Sema/SemaChecking.cpp
@@ -660,7 +660,12 @@ struct BuiltinDumpStructGenerator {
           return true;
       } else {
         Format += " ";
-        if (appendFormatSpecifier(FD->getType(), Format)) {
+        // refer SemaExpr.cpp 839-854
+        QualType PTy = {};
+        if(FD->getType()->isIntegralOrUnscopedEnumerationType()){
+          PTy = S.Context.isPromotableBitField(Field.get());
+        }
+        if (appendFormatSpecifier(!PTy.isNull() ? PTy : FD->getType(), Format)) {
           // We know how to print this field.
           Args.push_back(Field.get());
         } else {
